ComputerVision/CIFAR/utils/optim_utils.py

```diff
@@ -0,0 +1,237 @@
+        )
+    else:
+        lr_str += "."
+        lr_scheduler = main_lr_scheduler
+
+    print(f"{lr_str}")
+
+    return optimizer, lr_scheduler
+
+
+def compute_explicit_weight_decay_low_rank(
+    model, weight_decay, no_wd_bn_bias, apply_on_folded=True
+):
+    '''Computes weight-decay on the reconstructed weight-tensor for 
+        stochastic low-rank factorized layers.
+
+    Args:
+        model (torch.nn.Module): DNN model.
+        weight_decay (float): magnitude of weight-decay.
+        no_wd_bn_bias (bool): if True, we do not apply weight-decay 
+            on bn and bias.
+        apply_on_folded (bool): if True, we apply weight-decay by 
+            folding low-rank layers.
+    '''
+
+    loss_weight_decay = torch.tensor(0.0, requires_grad=True)
+    model_parameters = dict(model.named_parameters())
+    model_modules = dict(model.named_modules())
+    for name, param in model.named_parameters():
+        if not param.requires_grad:
+            continue
+        if len(param.shape) == 1:
+            # Catches Bias, and BN params
+            if no_wd_bn_bias:
+                # Skip weight-decay on BN and bias
+                continue
+            loss_weight_decay = (
+                loss_weight_decay + weight_decay * 0.5 * (param ** 2).sum()
+            )
+        else:
+            # Catches all the weight-tensors
+            if apply_on_folded:
+                if 'U_' in name:
+                    # We will fetch value of weight-decay computed and stored in the module
+                    name_split = name.split('.')
+                    module_name = '.'.join(name_split[:3])
+                    _weight_norm = model_modules[
+                        module_name
+                    ].dynamic_weight_norm
+                    loss_weight_decay = (
+                        loss_weight_decay + weight_decay * 0.5 * _weight_norm
+                    )
+                    # loss_weight_decay = (
+                    #     loss_weight_decay
+                    #     + weight_decay * 0.5 * (param ** 2).sum()
+                    # )
+                elif 'V_' in name:
+                    # We do not apply weight-decay on V component, we apply on W 
+                    # constructed by folding U and V.
+                    continue
+                    # loss_weight_decay = (
+                    #     loss_weight_decay
+                    #     + weight_decay * 0.5 * (param ** 2).sum()
+                    # )
+                elif 'sparse_conv' in name:
+                    loss_weight_decay = (
+                        loss_weight_decay
+                        + weight_decay * 0.5 * (param ** 2).sum()
+                    )
+                    # We do not apply weight-devay on sparse_conv component,
+                    # we apply on W constructed by folding UV and sparse(W).
+                    # continue
+                    # loss_weight_decay = loss_weight_decay + weight_decay*0.5*(param**2).sum()
```

Comment:
nit: remove unsed / commented code for cleaner parsing

Reply:
done